Fix inconsistencies in node contracts states

sameh-farouk opened the pull request:
## Description

- A defensive check has been added to ensure that node contracts verify the state of the associated rent contract before transitioning to the created or resumed state.
- When a rent contract attempts to resume a node contract, it now checks for any unsettled funds associated with that node contract.
- The deletion of rent contracts is now delayed until all associated node contracts (with incurred costs) are independently billed and deleted. This ensures that no active node contracts have reserved balances before proceeding with rent contract deletion.
- State transitions from gracePeriod(x) to gracePeriod(y) are now blocked to prevent improper state changes.

## Related Issues:

https://github.com/threefoldtech/tfchain/issues/1002

## Checklist:

Please delete options that are not relevant.

- [x] I have added tests to cover my changes.
- [X] My commits follow this [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) guide.

sameh-farouk committed:
fix: fix inconsistencies in node contracts states

renauter commented on substrate-node/pallets/pallet-smart-contract/src/billing.rs:
```diff
@@ -362,8 +362,14 @@ impl<T: Config> Pallet<T> {
         has_sufficient_fund: bool,
         current_block: u64,
     ) -> DispatchResultWithPostInfo {
+        // Add a defensive check to prevent a node contract on a rented node from transitioning to 'created' status without considering the state of the associated rent contract
+        let rent_is_suspended = matches!(contract.contract_type, types::ContractData::NodeContract(_))
+        && ActiveRentContractForNode::<T>::get(contract.get_node_id())
+            .and_then(|id| Contracts::<T>::get(&id))
+            .map(|c| matches!(c.state, types::ContractState::GracePeriod(_)))
+            .unwrap_or(false);
```
Could we move this block inside the match specific case ?
Something like:
~~~rust
types::ContractState::GracePeriod(_) if has_sufficient_fund => {
    // Manage transition from GracePeriod to Created

    // Add a defensive check to prevent a node contract on a rented node from transitioning to 'created' status without considering the state of the associated rent contract
        let node_contract_on_suspended_rent =
            matches!(contract.contract_type, types::ContractData::NodeContract(_))
                && ActiveRentContractForNode::<T>::get(contract.get_node_id())
                    .and_then(|id| Contracts::<T>::get(&id))
                    .map(|c| matches!(c.state, types::ContractState::GracePeriod(_)))
                    .unwrap_or(false);

    if node_contract_on_suspended_rent {
        log::debug!(
            "Node contract is on a rented node and should remain in grace period if rent contract it is on is still in grace period."
        );
        return Ok(().into());
    }

    ...
}
~~~

sameh-farouk replied:
This won't work correctly because it needs to match with remaining cases, specifically the case that manages the transition from GracePeriod to Deleted.

renauter replied:
Let s take following example:
If I have a node contract in grace which is on a rent contract, also in grace, (`rent_is_suspended == true`) and I have sufficient funds (`has_sufficient_fund == true`) I would fall in the `transition from GracePeriod to Deleted` case, correct ?
Not sure that is what we want to achieve, unless I miss something.

I though the idea was to avoid to switch a node contract on grace on a rent contract on grace to `Created` and that this transition should be done by handling at first the rent contract (propagating the state to the "child" node contracts by calling `synchronize_associated_node_contract_states()` )

sameh-farouk replied:
The node contract is associated with a workload, which in this case includes an IP address. The user has just add enough balance to cover the IP rent and NU consumption, but the rent contract is currently in a grace period, meaning the user has not yet paid for the node resources he is utilizing. As a result, the node contract should remain in the grace period as long as his rent contract suspended. During this time, the system should regularly check if the grace period has expired (if the user has reached the end of the 14-day grace period, allowing the contract to transition to deletion). If the grace period has not yet expired, the contract will remain in its current state.

sameh-farouk replied:
This allows the node contract to transition to the deleted state, thereby distributing any reserved balance accordingly. Only then, rent contract would follow in next cycle.

renauter replied:
Ok, I see now
Thank you for explaining

Could we just change variable name `rent_is_suspended` to something like `node_contract_on_suspended_rent` ?
Would be clearer that in case `== true` the input contract is a node contract and not a rent contract

sameh-farouk replied:
I prefer `rent_is_suspended` name, and I believe the context and the comment are clear. let's keep that, please.

renauter commented on substrate-node/pallets/pallet-smart-contract/src/billing.rs:
```diff
@@ -386,6 +392,26 @@ impl<T: Config> Pallet<T> {
+                    };
+                    if !can_be_deleted {
+                        log::debug!(
+                            "Grace period expired, but one or more associated node contracts have held user funds. \
```
Grace period on rented node expired, but one or more associated node contracts hold user funds.

renauter commented on substrate-node/pallets/pallet-smart-contract/src/billing.rs:
```diff
@@ -386,6 +392,26 @@ impl<T: Config> Pallet<T> {
+                    // Ensure associated node contracts have no reserved balance.
+                    // If none, proceed to move the contract to the 'deleted' state.
+                    // Otherwise, wait.
+                    let can_be_deleted = match contract.contract_type {
+                        types::ContractData::RentContract(_) => {
+                            let active_node_contracts = ActiveNodeContracts::<T>::get(contract.get_node_id());
+                            !active_node_contracts.iter().any(|id| {
+                                ContractPaymentState::<T>::get(id).map_or(false, |s| s.has_reserve())
+                            })
+                        }
+                        _ => true,
+                    };
```
Can we have more explicit naming here and insert `if` inside `match` ?
Like:
~~~rust
                    match contract.contract_type {
                        types::ContractData::RentContract(_) => {
                            let active_node_contracts = ActiveNodeContracts::<T>::get(contract.get_node_id());
                            let rent_has_node_contracts_with_reserve = 
                            active_node_contracts.iter().any(|id| {
                                ContractPaymentState::<T>::get(id).map_or(false, |s| s.has_reserve())
                            })

                            if rent_has_node_contracts_with_reserve {
                                log::debug!(
                                    "Grace period expired, but one or more associated node contracts have held user funds. \
                                    Rent contract deletion will be delayed until funds are distributed to beneficiaries."
                                );
                                return Ok(().into());
                            }
                        }
                        _ => (),
                    };
                    

~~~

renauter commented on substrate-node/pallets/pallet-smart-contract/src/billing.rs:
```diff
@@ -734,6 +760,14 @@ impl<T: Config> Pallet<T> {
                         Contracts::<T>::get(ctr_id).ok_or(Error::<T>::ContractNotExists)?;
+
+                    // Defensive check to prevent a node contract on a rented node from transitioning to 'created' status if there's an overdraft (unsettled public IP rent).
+                    if ContractPaymentState::<T>::get(ctr_id).map_or_else(|| false, |ps| ps.has_overdraft())
```
Just a doubt...
Why checking for overdraft here and checking for reserve at line 402 ?

sameh-farouk replied:
Handling different cases.
Here we check for overdraft on node contracts to prevent it from transitioning to 'created' status if there's an overdraft associated with it.

In line 402, we check for any reserved balance on node contracts to prevent the rent contract and its node contracts to transitioning to deleted state till held funds are distributed to beneficiaries.

sameh-farouk replied:
This allows the node contract to transition to the deleted state, thereby distributing any reserved balance accordingly. Only then, rent contract would follow in next cycle.

renauter commented on substrate-node/pallets/pallet-smart-contract/src/grid_contract.rs:
```diff
@@ -270,9 +270,12 @@ impl<T: Config> Pallet<T> {
-        let state = contract.state.clone();
-        Self::update_contract_state(&mut contract, &state)?;
-
+        // Why we call update_contract_state if state can't be changed?
+        // Obviously, we just depened on the update_contract_state for writing the contract to the storage, but we can do it here
+        // let state = contract.state.clone();
+        // Self::update_contract_state(&mut contract, &state)?;
+        Contracts::<T>::insert(&contract.contract_id, contract.clone());
```
Good catch
I also don t see the reason why `update_contract_state()` should be called here since `deployment_hash` and `deployment_data` have nothing to do with contract state and what is handled inside `update_contract_state()`.

sameh-farouk committed:
test: add tests for scenarios related to issue #1002

sameh-farouk committed:
refactor: small code refactor based on PR comment

github-actions[bot] committed:
chore: update benchmark `weights.rs` file for all pallets

sameh-farouk committed:
Merge remote-tracking branch 'origin' into development_fix_sync_contracts

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+        let node_id = 1;
+
+        // Set the TFT prices
+        TFTPriceModule::set_prices(RuntimeOrigin::signed(alice()), 50, 101).unwrap();
```
Seems TFT price is already set inside `prepare_dedicated_farm_and_node()`, and also inside `prepare_farm_and_node()` btw.
Seems it is duplicated in many tests of this file.
I guess we should get rid of these duplicated by removing this line in all tests where it is happening.

Also I am not sure why we input block `101` (even if no impact)

sameh-farouk replied:
Indeed. I removed it from the new tests.
Other tests can be reviewed and adjusted maybe in other PR

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+        // Transfer some balance to the owner of the contract to settle only the cost of IP rent (node contract)
+        // Case 1: See https://github.com/threefoldtech/tfchain/issues/1002
+        // Calculate the node contract cost
+        let (amount_due, discount_level) = calculate_tft_cost(node_contract_id, 2, 10);
```
Could we add a variable `let bob_twin_id = 2` just before for readability?
Also on next test

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            .should_call_bill_contract(node_contract_id, Ok(Pays::Yes.into()), 22);
+        run_to_block(22, Some(&mut pool_state));
+        
+        // Check if last event is Contract billed evenet
```
event

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+        // Check if last event is Contract billed evenet
+        let events = System::events();
+        log::debug!("Events: {:?}", events);
+        let contract_bill_event = types::ContractBill {
```
Could we rename `contract_bill_event` to `contract_bill` ?

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            .should_call_bill_contract(node_contract_id, Ok(Pays::Yes.into()), 22);
+        run_to_block(22, Some(&mut pool_state));
+        
+        // Check if last event is Contract billed evenet
```
event

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            .should_call_bill_contract(rent_contract_id, Ok(Pays::Yes.into()), 21);
+        run_to_block(21, Some(&mut pool_state));
+
+        // Check if evenets contains ContractGracePeriodStarted events for both contracts
```
events contain

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            2
+        );
+
+        // Transfer some balance to the owner of the contract to settel only the cost of IP rent (node contract)
```
settle

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+                ))
+                .count(),
+            2
+        );
+
+        // Transfer some balance to the owner of the contract to settel only the cost of IP rent (node contract)
+        // Case 2: see https://github.com/threefoldtech/tfchain/issues/1002
+        // Calculate the node contract cost
+        let (amount_due, discount_level) = calculate_tft_cost(node_contract_id, 2, 10);
+        log::debug!("Amount due: {}, discount level: {:?}", amount_due, discount_level);
+        Balances::transfer_allow_death(RuntimeOrigin::signed(charlie()), bob(), amount_due).unwrap();
+
+        pool_state
+            .write()
+            .should_call_bill_contract(node_contract_id, Ok(Pays::Yes.into()), 22);
+        run_to_block(22, Some(&mut pool_state));
+        
+        // Check if last event is Contract billed evenet
+        let events = System::events();
+        log::debug!("Events: {:?}", events);
+        let contract_bill_event = types::ContractBill {
+            contract_id: node_contract_id,
+            timestamp: 1628082132,
+            discount_level,
+            amount_billed: amount_due as u128,
+        };
+        assert_eq!(
+            events
+                .last()
+                .unwrap()
+                .event,
+            MockEvent::SmartContractModule(SmartContractEvent::ContractBilled(contract_bill_event))
+        );
+
+        // Node contract cost settled (IP address cost) but should remain in grace period due to suspened rent contract
+        // Because the cost of the node resources utlized by this node contract is billed to the rent contract
+        let node_contract = SmartContractModule::contracts(node_contract_id).unwrap();
+        assert_eq!(node_contract.state, types::ContractState::GracePeriod(21));
+    });
+}
```
I don t see any difference between this test and the previous one
Is it on purpose ?

sameh-farouk replied:
Actually, after writing the two scenarios, I intended to swap their order to match the sequence in the issue. However, something went wrong during the swap, and I ended up with duplicate tests. I have now committed the missing one again.
Thanks for bringing that to my attention

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            .should_call_bill_contract(rent_contract_id, Ok(Pays::Yes.into()), 21);
+        run_to_block(21, Some(&mut pool_state));
+
+        // Check if evenets contains ContractGracePeriodStarted events for both contracts
```
events contain

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+        // Case 3: see https://github.com/threefoldtech/tfchain/issues/1002
+        for i in 1..=10 {
+            let block_number = 21 + i * BillingFrequency::get();
+            pool_state.write().should_call_bill_contract(
+                rent_contract_id,
+                Ok(Pays::Yes.into()),
+                block_number,
+            );
+            run_to_block(block_number, Some(&mut pool_state));
+            pool_state.write().should_call_bill_contract(
+                node_contract_id,
+                Ok(Pays::Yes.into()),
+                block_number+1,
+            );
+            run_to_block(block_number+1, Some(&mut pool_state));
+        }
+        // Check if the events contains RewardDistributed event
+        let events = System::events();
+        log::debug!("Events: {:?}", events);
+        assert_eq!(
+            events
+                .iter()
+                .filter(|e| matches!(
+                    e.event,
+                    MockEvent::SmartContractModule(SmartContractEvent::RewardDistributed { .. })
+                ))
+                .count(),
+            1
+        );
+        // Check if the last event is NodeContractCanceled event
+        assert_eq!(
+            events
+                .iter()
+                .filter(|e| matches!(
+                    e.event,
+                    MockEvent::SmartContractModule(SmartContractEvent::NodeContractCanceled { .. })
+                ))
+                .count(),
+            1
+        );
```
Checking for last event could be done like this:

~~~rust
        assert_eq!(
            events
                .last()
                .unwrap()
                .event,
            MockEvent::SmartContractModule(SmartContractEvent::{
                    contract_id: node_contract_id,
                    node_id,
                    twin_id: bob_twin_id
                })
        );
~~~

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+                .iter()
+                .filter(|e| matches!(
+                    e.event,
+                    MockEvent::SmartContractModule(SmartContractEvent::RewardDistributed { .. })
+                ))
+                .count(),
+            1
+        );
+        // Check if the last event is NodeContractCanceled event
+        assert_eq!(
+            events
+                .iter()
+                .filter(|e| matches!(
+                    e.event,
+                    MockEvent::SmartContractModule(SmartContractEvent::NodeContractCanceled { .. })
+                ))
+                .count(),
+            1
+        );
+        // The rent contract should be canceled in the next cycle
+        let block_number: u64 = 131;
+        pool_state.write().should_call_bill_contract(
+            rent_contract_id,
+            Ok(Pays::Yes.into()),
+            block_number,
+        );
+        run_to_block(block_number, Some(&mut pool_state));
+
+        // Check if the last event is RentContractCanceled event
+        let events = System::events();
+        assert_eq!(
+            events
+                .iter()
+                .filter(|e| matches!(
+                    e.event,
+                    MockEvent::SmartContractModule(SmartContractEvent::RentContractCanceled { .. })
+                ))
+                .count(),
+            1
+        );
```
same as above for checking last event

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -2994,6 +2994,436 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            block_number,
+        );
+        run_to_block(block_number, Some(&mut pool_state));
+
```
we could add extra check:
~~~rust
        // Check if the events contain RewardDistributed event for both contracts
        let events = System::events();
        log::debug!("Events: {:?}", events);
        assert_eq!(
            events
                .iter()
                .filter(|e| matches!(
                    e.event,
                    MockEvent::SmartContractModule(SmartContractEvent::RewardDistributed { .. })
                ))
                .count(),
            2
        );
~~~

to check for both contracts rewards distribution or is it out of scope ?

sameh-farouk replied:
Yes, IMO it is not needed for this scenario

renauter reviewed (commented):
Tests well cover the cases related to issue https://github.com/threefoldtech/tfchain/issues/1002
Nice work !!

Added my comments

sameh-farouk committed:
chore: fix typo in log comment

sameh-farouk committed:
test: correct case 2

sameh-farouk committed:
refactor: minor test refactor

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            );
+            run_to_block(block_number+1, Some(&mut pool_state));
+        }
+        // Check if the events contains RewardDistributed event
```
contain

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+        let node_contract = SmartContractModule::contracts(node_contract_id).unwrap();
+        assert_eq!(node_contract.state, types::ContractState::GracePeriod(11));
+
+        // Transfer some balance to the owner of the contract to settle only the cost of node resources (rent contract)
```
... for 2 billing cycles

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            .write()
+            .should_call_bill_contract(rent_contract_id, Ok(Pays::Yes.into()), 11);
+        run_to_block(11, Some(&mut pool_state));
+
```
would have been nice to have some overdraft check here:
~~~rust
assert_ne!(rent_contract_overdraft, 0);
assert_eq!(rent_contract_reserve, spendable_balance);
~~~
or even:
~~~rust
assert_eq!(rent_contract_overdraft + rent_contract_reserve, rent_contract_cost_1_cycle);
~~~

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            .write()
+            .should_call_bill_contract(node_contract_id, Ok(Pays::Yes.into()), 12);
+        run_to_block(12, Some(&mut pool_state));
+
```
also here something like:
~~~rust
assert_ne!(node_contract_overdraft, 0);
assert_eq!(node_contract_reserve, 0);
assert_eq!(rent_contract_overdraft, node_contract_ip_cost_1_cycle);
~~~

to be able to understand better what is happening in "backstage"

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+        // Case 2: see https://github.com/threefoldtech/tfchain/issues/1002
+        // Calculate the rent contract cost
+        let charlie_twin_id = 3;
+        let (contract_cost, discount_level) = calculate_tft_cost(rent_contract_id, charlie_twin_id, 20);
```
rename `contract_cost` to `rent_contract_cost_2_cycles` ?
and `amount_due` to `rent_amount_due`

sameh-farouk replied:
I’ve decided to skip these changes because the current names reflect their purpose without adding unnecessary complexity. The additional qualifiers don’t seem essential as the context is clear.

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+
+        let node_contract = SmartContractModule::contracts(rent_contract_id).unwrap();
+        assert_eq!(node_contract.state, types::ContractState::Created);
+
```
~~~rust
assert_eq!(rent_contract_overdraft, 0);
assert_eq!(rent_contract_reserve, rent_contract_cost_2_cycles);
assert_eq!(balance_reserve_charlie, rent_contract_cost_2_cycles);
~~~

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            .should_call_bill_contract(rent_contract_id, Ok(Pays::Yes.into()), 21);
+        run_to_block(21, Some(&mut pool_state));
+
+        let node_contract = SmartContractModule::contracts(rent_contract_id).unwrap();
```
~~~rust
let rent_contract = ...
~~~

renauter commented on substrate-node/pallets/pallet-smart-contract/src/tests.rs:
```diff
@@ -3033,6 +3033,399 @@ fn test_rent_contract_and_node_contract_canceled_when_node_is_deleted_works() {
+            .should_call_bill_contract(node_contract_id, Ok(Pays::Yes.into()), 12);
+        run_to_block(12, Some(&mut pool_state));
+
+        let node_contract = SmartContractModule::contracts(rent_contract_id).unwrap();
```
~~~rust
let rent_contract = ...
~~~

renauter reviewed (commented):
Reviewed test case 2
Test flow is great !
See my suggestions for improvement

sameh-farouk committed:
chore: fix typos

sameh-farouk commented:
@renauter  Thank you for your thoughtful review. I’ve addressed most of your suggestions.

I also appreciate the suggestion to add more assertions, but I believe these extra checks go beyond the scope of this test case. The existing assertions already verify the core functionality being fixed, and adding more may detract from the specific focus of the test.

github-actions[bot] committed:
chore: update benchmark `weights.rs` file for all pallets
